flask_monitoringdashboard/core/cache.py

```diff
@@ -0,0 +1,111 @@
+"""
+    Contains the in memory cache used to increase the FMD performance.
+"""
+import datetime
+from multiprocessing import Lock
+
+from flask_monitoringdashboard.core.rules import get_rules
+from flask_monitoringdashboard.database import session_scope
+from flask_monitoringdashboard.database.endpoint import get_last_requested, get_endpoints_hits, get_endpoint_averages, \
+    update_last_requested
+
+memory_cache = {}
+mutex = Lock()
+
+
+class EndpointInfo(object):
+    """
+    Info about an endpoint that is stored in the memory cache.
+    """
+    def __init__(self, last_requested=None, average_duration=None, hits=None):
+        self.last_requested = last_requested
+        self.average_duration = average_duration if average_duration else 0
+        self.hits = hits if hits else 0
+
+    def set_last_requested(self, last_requested):
+        with mutex:
+            self.last_requested = last_requested
+
+    def set_duration(self, duration):
+        with mutex:
+            self.average_duration = (self.average_duration * self.hits + duration)/float(self.hits + 1)
+            self.hits += 1
+
+    def get_duration(self):
+        with mutex:
+            return self.average_duration
+
+
+def display_cache():
+    """
+    Debug purposes.
+    """
+    global memory_cache
+    for k in memory_cache.keys():
+        print('%s : last=%s, avg=%f, hits=%d' % (k, memory_cache[k].last_requested,
+                                                 memory_cache[k].average_duration, memory_cache[k].hits))
+
+
+def init_cache():
+    """
+    This should be added to the list of functions that are executed before the first request.
+    It initializes the in memory cache from the db
+    """
+    global memory_cache
+    with session_scope() as db_session:
+        last_req_dict = dict(get_last_requested(db_session))
+        hits_dict = dict(get_endpoints_hits(db_session))
+        averages_dict = dict(get_endpoint_averages(db_session))
+        for rule in get_rules():
+            memory_cache[rule.endpoint] = EndpointInfo(last_requested=last_req_dict.get(rule.endpoint),
+                                                       average_duration=averages_dict.get(rule.endpoint),
```

Comment:
which average is this one? 
is it the alltime? last 7 days? or it does not matter? 
would it make sense to add a corresponding comment in the EndpointInfo?

Reply:
It's alltime. It's used to determine the sleeping time of the outlier thread. But agree, will add a comment